Test harness documentation improvement (#935)
medic/cht-docs: content/en/apps/tutorials/application-tests.md

## Prerequisites

Complete the following tutorials:
  - [Building App Forms]({{< ref "apps/tutorials/app-forms" >}})
  - [Building A Simple Task]({{< ref "apps/tutorials/tasks-1" >}})
  - [Building Target Widgets]({{< ref "apps/tutorials/targets" >}})
  - [Building Contact Summary]({{< ref "apps/tutorials/contact-summary" >}})

## Importance of testing your application
---
Testing your CHT application is important as it ensures you are consistently maintaining your application and defining implementation requirements. Testing also helps the user make better architectural decisions, optimizes the forms, tasks and other components of the application. Although it may seem tedious, testing your application ensures that you quickly discover issues that are introduced with changes or newer commits.
## CHT Application Testing
---

> CHT applications are greatly configurable. Depending on the number and complexity of app components, it can take a lot of time and effort to test the components  manually. Some components, such as tasks, behave differently over time and are particularly challenging to test. As the project evolves, the configuration is often updated with new components and changes are made to the existing components. After each change, you need to test not only the new components, but also the old ones to make sure that the app works as expected. To facilitate the testing process, app builders are encouraged to write automated tests for their app using [cht-conf-test-harness](http://docs.communityhealthtoolkit.org/cht-conf-test-harness/).

Reviewer: Extra space.
```suggestion
CHT applications are greatly configurable. Depending on the number and complexity of app components, it can take a lot of time and effort to test the components manually. Some components, such as tasks, behave differently over time and are particularly challenging to test. As the project evolves, the configuration is often updated with new components and changes are made to the existing components. After each change, you need to test not only the new components, but also the old ones to make sure that the app works as expected. To facilitate the testing process, app builders are encouraged to write automated tests for their app using [cht-conf-test-harness](http://docs.communityhealthtoolkit.org/cht-conf-test-harness/).
```

Because it is complicated to test with a real application, `cht-conf-test-harness` (also simply referred as `harness`), provides a platform that simulates the CHT application instance.

Using `cht-conf-test-harness`, you can write tests and run them with [Mocha](https://mochajs.org/) testing framework to test the behavior of different components in a CHT application. You can also use [chai library](https://www.chaijs.com/) or other assertion libraries.

## Preparation

Writing tests for CHT apps requires a good understanding of the project workflows and requirements. To test using the harness, there are a few things you need to set up:
1. From the previous tutorials, you should have a [functioning CHT instance with `cht-conf` installed locally]({{< ref "apps/tutorials/local-setup" >}}) and a [project folder set up]({{< ref "apps/tutorials/local-setup#3-create-and-upload-a-blank-project" >}}). 

> `cht-conf` which is short for CHT app configurer, is a command-line interface tool used to manage and configure your CHT apps. It is used for form conversion, form validation and other actions which can be found by running `cht --help`.

Reviewer: > form conversion, form validation and other actions

Too much focus on just the forms?

Author: @binokaryg , I have incorporated this suggestion.

2. Ensure your `package.json` file has the required libraries. A `package.json` file is used to record important metadata about a project and defines functional attributes that npm uses to install dependencies and run scripts. This file should be at the root of your project folder. 
If your `package.json` file does not already have them, add `cht-conf-test-harness`, `chai`, and `mocha` by running this in your command-line:
    ```shell

>     npm install --save cht-conf-test-harness chai mocha

Reviewer 2: ```suggestion
    npm install --save-dev cht-conf-test-harness chai mocha
```
More accurate to use [devDependencies](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#devdependencies) here

    ```
    After adding these scripts, you are able to run the tests by running one of these commands from the command-line:

    |Command|Description|
    |---|---|
    |`npm run test` |Compiles the app settings, converts the forms,then runs the tests (preferred solution).|
    |`npm run unittest`| Only runs the unit tests.|

     You can also run the test on a specific form by defining a test under the scripts section.
     
     For example, to run the test on the `death.spec.js` file, define the test as `death-form-unit-test: "mocha test/forms/death.spec.js --timeout 20000"`. Then run:
     ```shell
        npm run death-form-unit-test
     ```   
   

>      Npm also supports a shorter alias as `npm test`.There are recommended conventions on how to arrange tests by splitting into unit tests and integration tests so that each of them can be run independently. This can be best illustrated by [this project code](https://github.com/medic/cht-pih-malawi-app).

Reviewer: It does not look right to write `npm` as `Npm`. Also, space needed after period.
```suggestion
     npm also supports a shorter alias as `npm test`. There are recommended conventions on how to arrange tests by splitting into unit tests and integration tests so that each of them can be run independently. This can be best illustrated by [this project code](https://github.com/medic/cht-pih-malawi-app).
```

   |Command|Description|
   |---|---|
   |`npm run test`|Compiles the app settings, converts the forms, then runs the tests (preferred solution).
   |`npm run unittest`|Only runs the unit tests.|

>    You can also run the test on a specific form by defining a test under the scripts section. For example to run the test on the death.spec.js file:

Reviewer: ```suggestion
   You can also run the test on a specific form by defining a test under the scripts section. For example to run the test on the `death.spec.js` file:
```

      `death-form-unit-test: "mocha test/forms/death.spec.js --timeout 20000"`
   Then run:
      `npm run death-form-unit-test`
   |Command|Description|
   |---|---|
   |`npm run test`|Compiles the app settings, converts the forms, then runs the tests (preferred solution).
   |`npm run unittest`|Only runs the tests.|

4. Create a folder in the project root where you keep the tests. You can name the folder yourself. For convenience, here it is named as `test`.
{{< figure src="test-folder-placement.png" link="test-folder-placement.png" class="col-9 col-lg-12" >}}

5. Create a file `harness.defaults.json` at the root of your project. This is the default configuration file for the harness. Here you can specify the default user, preloaded contacts and reports, and add other settings. Here's an [example](https://github.com/medic/cht-conf-test-harness/blob/master/harness.defaults.json.example) file to get you started. You can read more about it [here](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/global.html#HarnessInputs).
{{< figure src="harness-json-file-placement.png" link="harness-json-file-placement.png" class="col-9 col-lg-12" >}}

## Writing a test for CHT App

> Start by adding a file where a group of related tests are written. For example, in the default config, all tests for the pregnancy form are in the [pregnancy.spec.js](https://github.com/medic/cht-core/blob/master/config/default/test/forms/pregnancy.spec.js) file. The commmon naming pattern for CHT applications test files is filename.spec.js. In Javascript testing, spec is short for specifications which refers to technical details of a given application, that must be fulfilled.

Reviewer: ```suggestion
Start by adding a file where a group of related tests are written. For example, in the default config, all tests for the pregnancy form are in the [pregnancy.spec.js](https://github.com/medic/cht-core/blob/master/config/default/test/forms/pregnancy.spec.js) file. The common naming pattern for CHT applications test files is `filename.spec.js`. In JavaScript testing, `spec` is short for specifications which refers to technical details of a given application, that must be fulfilled.
```

      `death-form-unit-test: "mocha test/forms/death.spec.js --timeout 20000"`
   Then run:
      `npm run death-form-unit-test`
   |Command|Description|
   |---|---|

>    |`npm run test`|Compiles the app settings, converts the forms, then runs the tests (preferred solution).

Author: @eljhkrr I have added this suggestion. Thanks

>    |`npm run unittest`|Only runs the tests.|

Reviewer 2: It's important to mention these are recommended conventions on how to arrange tests by splitting into unit tests and integration tests so they can each be run independently. This can be best illustrated by sharing an example app such as https://github.com/medic/cht-pih-malawi-app

Author: @eljhkrr Done!

```

> In the example above, we define values `TestRunner` and `harness` .You get an instance of `harness` and the same instance is used throughout the test file.

Reviewer: ```suggestion
In the example above, we define values `TestRunner` and `harness`. You get an instance of `harness` and the same instance is used throughout the test file.
```

```

> In the example above, we define values `TestRunner` and `harness`. You get an instance of `harness` and the same instance is used throughout the test file.

Reviewer 2: ```suggestion
In the example above, we define variables `TestRunner` and `harness`. `TestRunner` imports the Harness class and `harness` creates an instance of the class that is used throughout the test file.
```
